src/checkers/flymake-collection-mypy.el

```diff
@@ -26,34 +26,98 @@
 
 ;;; Code:
 
+(require 'project)
 (require 'flymake)
 (require 'flymake-collection)
 
 (eval-when-compile
   (require 'flymake-collection-define))
 
+(defcustom flymake-collection-mypy-args
+  '("--show-error-codes")
+  "Command line arguments always passed to `flymake-collection-mypy'."
+  :type '(repeat (string :tag "Arg"))
+  :group 'flymake-collection)
+
+(defcustom flymake-collection-mypy-project-root
+  '(("mypy.ini" "project.toml" "setup.cfg") project-root default-directory)
+  "Method to set project root."
+  :type '(repeat :tag "Run mypy from the first choice that succeeds"
+                 (choice (const :tag "The buffer default-directory" default-directory)
+                         (const :tag "The current project root" project-root)
+                         (directory :tag "A specific directory")
+                         (repeat :tag "The first ancestor directory containing"
+                                 :value ("mypy.ini" "pyproject.toml" "setup.cfg")
+                                 (string :tag "File name"))))
+  :group 'flymake-collection
+  :safe 'listp)
+
+(defun flymake-collection-mypy--locate-dominating-files (buffer files)
+  "Find ancestor directory of `BUFFER' containing any of `FILES'."
+  (let* ((start (if-let ((file (buffer-file-name buffer)))
+                    (file-name-directory file)
+                  (buffer-local-value 'default-directory buffer)))
+         (regex (mapconcat 'regexp-quote files "\\|"))
+         (root (locate-dominating-file
+                start (lambda (dir) (directory-files dir nil regex t)))))
+    root))
+
+(defun flymake-collection-mypy--default-directory (buffer)
+  "Find a directory from which to run mypy to check `BUFFER'.
+Try each method specified in `flymake-collection-mypy-project-root' in
+order and returns the first non-nil result."
+  (cl-dolist (spec flymake-collection-mypy-project-root)
+    (when-let
+        ((res (cond
+               ((eq spec 'default-directory)
+                (buffer-local-value 'default-directory buffer))
+               ((eq spec 'project-root)
```

Comment:
Thanks for adding a project root check. I'm still not sure whether a recursive search up through the directory tree is the best (or at the very least a good default) choice here. Especially if this is something that's being edited across TRAMP or something else which adds a level of latency. That said I also can't think of a very good alternative either, this should really be something the tools themselves can pick up on IMO, not the packages that invoke those rules. That said they don't, so we should :upside_down_face:. I'm happy to approve this section. At some point it'd be nice if we could refactor it into something general purpose enough it could be used with the other python linters (they all have a habit of depending on this).